lib/internal/fs/watch/linux.js

```diff
@@ -0,0 +1,260 @@
+'use strict';
+
+const {
+  SafeMap,
+  StringPrototypeStartsWith,
+  Symbol,
+  SymbolAsyncIterator,
+} = primordials;
+
+const {
+  codes: {
+    ERR_FEATURE_UNAVAILABLE_ON_PLATFORM,
+    ERR_INVALID_ARG_VALUE,
+  },
+} = require('internal/errors');
+const { getValidatedPath } = require('internal/fs/utils');
+const { kEmptyObject } = require('internal/util');
+const { validateObject, validateBoolean, validateAbortSignal } = require('internal/validators');
+const { EventEmitter, once } = require('events');
+const path = require('path');
+
+const kFSWatchStart = Symbol('kFSWatchStart');
+
+let internalSync;
+let internalPromises;
+
+function lazyLoadFsPromises() {
+  internalPromises ??= require('fs/promises');
+  return internalPromises;
+}
+
+function lazyLoadFsSync() {
+  internalSync ??= require('fs');
+  return internalSync;
+}
+
+async function traverse(dir, files = new SafeMap()) {
+  const { opendir } = lazyLoadFsPromises();
+
+  const filenames = await opendir(dir);
+
+  for await (const file of filenames) {
+    const f = path.join(dir, file.name);
+
+    files.set(f, file);
+
+    if (file.isDirectory()) {
+      await traverse(f, files);
+    }
+  }
+
+  return files;
+}
+
+class FSWatcher extends EventEmitter {
+  #options = null;
+  #closed = false;
+  #files = new SafeMap();
+  #rootPath = path.resolve();
+
+  constructor(options = kEmptyObject) {
+    super();
+
+    validateObject(options, 'options');
+
```

Comment:
```suggestion
  const { persistent, recursive, signal, encoding } = options;
```

Reply:
Would it be preferrable to do the destructuring before all the checks to avoid repeating `options.` in each if block?